text/0000-unicode-idents.md

```diff
@@ -0,0 +1,161 @@
+- Feature Name: unicode_idents
+- Start Date: 2018-06-03
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow non-ASCII letters (such as accented characters, Cyrillic, Greek, Kanji, etc.) in Rust identifiers.
+
+# Motivation
+[motivation]: #motivation
+
+Rust is written by many people who are not fluent in the English language. Using identifiers in ones native language eases writing and reading code for these developers.
+
+The rationale from [PEP 3131] nicely explains it:
+
+> ~~Python~~ *Rust* code is written by many people in the world who are not familiar with the English language, or even well-acquainted with the Latin writing system. Such developers often desire to define classes and functions with names in their native languages, rather than having to come up with an (often incorrect) English translation of the concept they want to name. By using identifiers in their native language, code clarity and maintainability of the code among speakers of that language improves.
+> 
+> For some languages, common transliteration systems exist (in particular, for the Latin-based writing systems). For other languages, users have larger difficulties to use Latin to write their native words.
+
+Additionally some math oriented projects may want to use identifiers closely resembling mathematical writing.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Identifiers include variable names, function and trait names and module names. They start with a letter or an underscore and may be followed by more letters, digits and some connecting punctuation.
+
+Examples of valid identifiers are:
+
+* ASCII letters and digits: `image_width`, `line2`, `Photo`, `el_tren`, `_unused`
+* words containing accented characters: `garçon`, `hühnervögel`
+* identifiers in other scripts: `Москва`, `東京`, ...
+
+Examples of invalid identifiers are:
+
+* Keywords: `impl`, `fn`, `_` (underscore), ...
+* Identifiers starting with numbers or containing "non letters": `42_the_answer`, `third√of7`, `◆◆◆`, ...
```

Comment:
I would actually really like it if the √ symbol would be allowed as well (e.g. to white list it). In academic formula context this can make formulas in code even closer to the formulas from papers and thus easier to read in the academic context (I've actively missed this symbol when doing the same with Python 3). As an example I take the right top most formal from [this picture](https://api.ning.com/files/GZs8qdvnOuG*E1a-rRsbP6-a*n2rRt05URyStNHbV-X8nkYX8qFmcqM43jy4Cco-uLExBO7wFJwqOXVKe24QXS2n7Y1dywu5/Capture.PNG): 
```rust
let √ = f64::sqrt;
let σ² = σ.powi(2);
let exp = f64::exp;
let result = 1/√(2*π*σ²) * exp(-(1/2) * ((x - x_abs)/σ))
```
instead of
```rust
let sqrt = f64::sqrt;
let σ² = σ.powi(2);
let exp = f64::exp;
let result = 1/sqrt(2*π*σ²) * exp(-(1/2) * ((x - x_abs)/σ))
```

Reply:
> I would actually really like it if the √ symbol would be allowed as well

I am sympathetic to this idea but would really want to avoid to diverge from UAX31 for a single character (or a few). I would prefer if someone created a "custom operators" RFC that allowed to define [additional operators](https://www.unicode.org/charts/PDF/U2200.pdf). This would enable the `√` and many more. The intuition for an identifier should still be "a letter followed by zero or more letters or digits".